### DISCUSSION TITLE
How to convert a .p12 certificate to use with Google Play App Signing?

### DISCUSSION THREAD

**Discussion by RossD20Studios** [airsdk/Adobe-Runtime-Support#3405] [Q&A] [answered]
Fellow AIR Developers - 

Beginning sometime in 2021, Google introduced a new mechanism for signing apps to be distributed on Google Play Store. For newer apps being created today, Google now requires the .AAB format and the use of Google App Signing.

However, many of us have applications created prior to 2021 that were signed with .p12 certificates that were created either through ADT or Adobe Animate.

Moving forward, Google highly recommends use of their Google App Signing, and this is the only way you can distribute apps via .AAB format (which is particularly desirable if you wish to deploy a single upload for both mobile and Google Games for PC). Note: You can still submit .APKs signed with .p12 in two separate tracks (for mobile and PC), but you cannot use the AAB format (which is the recommended deployment going into the future) without Google App Signing.

I have attempted to follow the instructions provided by @marchbold posted here on airsdk.dev:
[https://airsdk.dev/docs/tutorials/platform/android/packaging-android-app-bundles](https://airsdk.dev/docs/tutorials/platform/android/packaging-android-app-bundles)

Michael's instructions also align with those of this Medium article by Google Expert Joe Birch here:
[https://medium.com/google-developer-experts/exploring-google-play-app-signing-b4d296f4ee9](https://medium.com/google-developer-experts/exploring-google-play-app-signing-b4d296f4ee9)

However, it seems that the options available in Google Play Console have changed since these articles were written. Here is what I have observed in the current version of Google Play console:

I am able to follow the first step and use keytool to obtain the certificate alias from the .p12:
![image](https://github.com/user-attachments/assets/c657f80f-1ea8-459c-a913-ec2d9fbd6293)

However, the second step calls for "Next, start the process of signing up to "Play App Signing" in the Play console for your application and use the Export and Upload a key from java keystore method."
![image](https://github.com/user-attachments/assets/54bf10cc-e73b-4dfa-bec4-3c1272ffc0a2)

My Google Play Console DOES NOT present these options. Instead, mine looks like this:
![image](https://github.com/user-attachments/assets/37f0b8c5-3ba6-469c-8412-4163463dcc28)

Therefore, I am unable to complete the final step of the conversion because I do not have the encryptionkey from Google Play Console to replace the XXXXXXXX in the code below:
![image](https://github.com/user-attachments/assets/0ddb5880-0196-46f1-9334-abbbee2066c6)

Have any other developers been able to successfully convert their .p12 app signing mechanism to the newer Google Play App Signing? If so, please share your steps here - it would be greatly appreciated!

**Comment by marchbold** [comment]
@RossD20Studios The description has changed slightly but that first option is still what you need to do. Basically you need to use the pepk tool to upload your old certificate and ensure you follow step 5 there for "increased security". Step 5 is where you create the upload certificate that you'll use to sign the outputs from AIR. 

The pepk tool looks like it has changed slightly in that instead of providing a key as text you download it to a file.

**Comment by jeuxdaily** [comment]
hi
first download java 20
second download pepk.jar and encryption_public_key.pem then put it in folder with certificate
third in command line cmd open java 20 folder
and write this code
good luck

**Comment by RossD20Studios** [comment]
Thanks @marchbold - I need just a little more clarity to understand how exactly I should modify the code Google has provided. For clarity, there appear to be two parts to Google App Signing:
1) App Signing Key (which is what the .p12 currently does)
2) Upload Key (which Google indicates is optional for added security)

Starting with the first part, App Signing Key:

<img width="1011" alt="Screenshot 2024-08-20 at 4 54 55 PM" src="https://github.com/user-attachments/assets/aec68256-1140-492c-95a8-0f83cc49d712">
In the screenshot above, it says "Replace the arguments, and enter your keystore and key passwords when prompted."

Let's go over each one:
**foo.keystore:** Should this be the filename of my .p12 certificate or do I need to first convert my .p12 to a Java Keystore? That's where I got lost the first time.
**foo (as in the alias):** Following your document, I assume this will simply be "1", correct?
**encrypted_private_key_path:** This appears to be the file I will upload for this step. So should this just be "encrypted_certificate.zip" as in your example?
**/path/to/encryption_public_key.pem:** This is simply the public key file I download from Google console, right? So I can just go ahead and place that in the java folder as @jeuxdaily mentioned and run the complete script.

Assuming all that's correct, I should get an output of "encrypted_certificate.zip" and I upload this to complete the "App Signing Key" portion of Google Play Signing. Do I have it right?

Now, onto the Upload Key:
<img width="966" alt="Screenshot 2024-08-20 at 4 55 00 PM" src="https://github.com/user-attachments/assets/6a831ccb-2dbe-465b-a54d-e7de1e1a2a40">

Google says you first need to create a new upload key, steps here:
<img width="758" alt="Screenshot 2024-08-20 at 5 07 10 PM" src="https://github.com/user-attachments/assets/69c7d36c-2485-4237-908b-dbbf9cbd1477">

Creation of the key involves using Android studio, as linked here: https://developer.android.com/studio/publish/app-signing#generate-key

Replace the arguments in bold:
**upload-keystore.jks**: This appears to be the file generated by Android studio creating the upload key
**upload:** This is simply the alias used when creating the key, which, if following the docs above, is literally the word "upload"
**upload_certificate.pem:** I'm not sure why this is bold (unless you want to change the filename) but I take it that this is the file generated that I would then upload to Google Play Console

**So, a couple more questions on the upload key:**
1) How does the App Signing Certificate relate to the upload key? I am assuming that when I build my application with AIR (ADT or IntelliJ in my case) that I will still use my existing .p12 certificate to sign the app. Because I will have uploaded my certificate in step 1 (App Signing Key), Google should accept my app signing since the keys will match, correct?

2) Why would I need to keep my upload_certificate.pem locally for the Upload Key? Isn't the upload key only used for actually uploading the build (in which case, it's already stored in Play Console, so I assume ADT/IntelliJ does not need to sign the app with it, correct?)

Thanks again!

**Comment by RossD20Studios** [comment, marked as answer]
The procedures I posted [here](https://github.com/airsdk/Adobe-Runtime-Support/discussions/3405#discussioncomment-10401068) yesterday are correct. After some additional experimentation, I am able to answer my own remaining questions about using .p12 with Google Play App Signing:

1) **How does the App Signing Certificate relate to the upload key?** 

**Answer:** 
If you don't follow the optional step 5 to create an upload certificate, then you can continue to sign your application with your existing .p12 file after you've uploaded the signature to Google following the [steps above](https://github.com/airsdk/Adobe-Runtime-Support/discussions/3405#discussioncomment-10401068) for the App Signing key. 

If you do create an upload certificate, then you will need to sign your AIR application using your new upload key. Of course, if you followed the steps to the letter (like I did) and used Android Studio to create a JSK (JavaKeyStore) then you're going to need to convert that JKS file back into a .p12 so you can sign with ADT/IntelliJ again.

To do that, you'll need to use the keytool (also located in the Java Home/bin folder) and use the following command:
```
keytool \
    -importkeystore              \
    -srckeystore    KEYSTORE.jks \
    -destkeystore   KEYSTORE.p12 \
    -srcstoretype   JKS          \
    -deststoretype  PKCS12       \
    -srcstorepass   mysecret     \
    -deststorepass  mysecret     \
    -srcalias       myalias      \
    -destalias      myalias      \
    -srckeypass     mykeypass    \
    -destkeypass    mykeypass    \
    -noprompt
```

It should also be possible to simply convert your existing .p12 into an upload certificate as well (as @marchbold had suggested) by using this:

`keytool -export -rfc -keystore MYP12FILENAME.p12 -alias 1 -file upload_certificate.pem`

In this case, your App Signing Key and your Upload Keys would both be identical. I can't confirm whether or not Google allows the keys to match, but once you have uploaded your certificate, you can't change your upload key without special request from Google. So, make sure you decide how you want your key managed before uploading any files to Google Play Console.

Finally, to answer the question "How does App Signing and Upload Key relate to each other?" - The App Signing key is what is used to verify that your app is identical signature for app update purposes. The App Signing key NEVER changes for all of time, and that's why it's necessary to upload the signature from the .p12 you originally used to upload the app. (**Side Note:** Google does allow a way to create a new app signing key, but you still need the old one for users below Android 13 and you'd then have to manage multiple build uploads going forward for each release - which gets really messy.)

With the "upload key", Google now allows developers a way to change the key they sign their app with when uploading new builds (by special request). This way, if you lose your key or need to change it for some other reason, you can. Essentially, Google uses the upload key as the way of authenticating that you are uploading your own apps and then, internally, it re-signs the apps with your original App Signing Key signature for you (since that gets stored in your Google Play Console).

2) **Why would I need to keep my upload_certificate.pem locally for the Upload Key?**

**Answer:** 
If you create a new upload certificate, then you need this certificate stored locally (in the .p12 format, which ADT/IntelliJ both use for signing AIR apps) to sign your new builds going forward. If you try to sign with your original App-Singing .p12 file, then Google will warn you that your signature must match the one provided in your upload certificate.

**Reply by marchbold** [reply]
@RossD20Studios Apologies I have been dealing with a few family issues so missed this one, but glad you managed to work through it. 

Your interpretation of the differences between the certificates is spot on, but I can confirm these certificates can be the same. In fact, the default if you don't go through the advanced process of generating an upload certificate, is that they are the same.

- Upload: used to verify upload builds, can be changed, use locally for building and debugging 
- App: used to resign your build for distribution to clients via the store, cannot* be changed, once uploaded to the store you don't really have to worry about it. 

The key advantage to this process for AIR developers is when you have an old 1024 bit certificate which can no longer be used by Java/AIR to sign applications, you can let Google handle this certificate and instead use a new 2048 bit "upload" certificate for signing applications and local debugging.

Note:  if you are using services that verify your application via the signature, these may fail when testing locally with the "upload" certificate. Some services allow multiple signatures to be added so just make sure you add the signature for your upload certificate (along with your app certificate) so you can test local builds.

Lastly: Note this process is specific to the Google Play Store. Other stores have similar services now but may not operate the same. 

If I get some time I'll update the airsdk site.

**Comment by MalacTheLittle** [comment]
[Windows]
I got my head blown with conversion of .p12 to private key... no matter which jdk version I used I got following error:

```
Error: Unable to export or encrypt the private key
java.security.NoSuchAlgorithmException: Cannot find any provider supporting RSA/NONE/OAEPWithSHA1AndMGF1Padding
        at java.base/javax.crypto.Cipher.getInstance(Cipher.java:573)
        at com.google.wireless.android.vending.developer.signing.tools.extern.export.ExportEncryptedPrivateKeyTool.encryptPrivateKeyWithCkmRsaAesKeyWrapEncryption(ExportEncryptedPrivateKeyTool.java:284)
        at com.google.wireless.android.vending.developer.signing.tools.extern.export.ExportEncryptedPrivateKeyTool.run(ExportEncryptedPrivateKeyTool.java:213)
        at com.google.wireless.android.vending.developer.signing.tools.extern.export.ExportEncryptedPrivateKeyTool.main(ExportEncryptedPrivateKeyTool.java:165)
```

One question of my colleague saved the day: "Which version of jdk you're using, Oracle or Open jdk?"
In short, to make it work you need to use [**OpenJDK**](https://jdk.java.net/archive/)... I used version JDK 22.0.1

And here's complete command prompt command:
```
"C:\Program Files\Java\jdk-22.0.1\bin\java.exe" -jar "E:\key\pepk.jar" --keystore="E:\key\cert.p12" --alias=1 --output="E:\key\encrypted_private_key.pem" --rsa-aes-encryption --encryption-key-path="E:\key\encryption_public_key.pem"
```
(it will prompt you to enter certificate password twice)
```
Enter password for store 'cert.p12':
Enter password for key '1':
```

**Comment by 2jfw** [comment]
I came across a similar issue to the certificate creation/conversions and would like to share what I encountered and how I solved it:

_The screenshots and infos are based on a completely new app, which has yet not been uploaded (neither releases nor any tests)._

1) My view for "**App Signing**" in the Google Play Console looked even different than as it was posted here (and, as already mentioned here, it also looked different than in the tutorial on airsdk.dev). I had the following view:

![2024-10-06 14_42_01-App signing _ Steamshot Beta – Mozilla Firefox](https://github.com/user-attachments/assets/0c88f8be-4c29-4165-851b-a5eab48fbc9b)

There is now a new stand-alone menu item "**App integrity**" which looks like this:
![2024-10-06 14_43_45-App integrity _ Steamshot Beta – Mozilla Firefox](https://github.com/user-attachments/assets/94e6772f-e2e0-4df0-ad8a-292365cb31dc)

**However, there is no option to select the app signing key!**

I found it under "**Testing / Open testing / Releases / Edit Releases**":
![Open Testing view](https://github.com/user-attachments/assets/5a45e344-b9b1-4b80-9c52-56f3fc931a46)

![2024-10-06 14_57_50-Prepare release _ Steamshot Beta – Mozilla Firefox](https://github.com/user-attachments/assets/d2bbc980-3444-41aa-b702-e174f1c1b172)

![App singning preferences Popup](https://github.com/user-attachments/assets/e150f91c-cc18-4d47-b878-bf4e2caa7cd7)


In the following steps, as described here or in the tutorial, I encountered the following issue as I tried to convert the `encryption_public_key.pem`:


`java -jar pepk.jar --keystore=cert.p12 --alias=1 --output=output.zip --include-cert --rsa-aes-encryption --encryption-key-path=encryption_public_key.pem`

caused:

```
Eception:
Exception in thread "main" java.lang.UnsupportedClassVersionError: com/google/wireless/android/vending/developer/signing/tools/extern/export/ExportEncryptedPrivateKeyTool has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0
        at java.lang.ClassLoader.defineClass1(Native Method)
        at java.lang.ClassLoader.defineClass(Unknown Source)
        at java.security.SecureClassLoader.defineClass(Unknown Source)
        at java.net.URLClassLoader.defineClass(Unknown Source)
        at java.net.URLClassLoader.access$100(Unknown Source)
        at java.net.URLClassLoader$1.run(Unknown Source)
        at java.net.URLClassLoader$1.run(Unknown Source)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.net.URLClassLoader.findClass(Unknown Source)
        at java.lang.ClassLoader.loadClass(Unknown Source)
        at sun.misc.Launcher$AppClassLoader.loadClass(Unknown Source)
        at java.lang.ClassLoader.loadClass(Unknown Source)
        at sun.launcher.LauncherHelper.checkAndLoadMain(Unknown Source)

```

I had to explicitly use the java.exe from (in my case) OpenJDK 21 like the following:

`"C:\Program Files\Java\jdk-21.0.2\bin\java" -jar pepk.jar --keystore=E:\certs\cert.p12 --alias=1 --output=output.zip --include-cert --rsa-aes-encryption --encryption-key-path=E:\certs\encryption_public_key.pem`

`pepk.jar` **must** be in folder "C:\Program Files\Java\jdk-21.0.2\bin" (this is also the folder where the output.zip can then be found).

`output.zip` contains:
- encryptedPrivateKey
- certificate.pem

I'm now in the process of creating the `upload_certificate.pem` ...
